### DISCUSSION TITLE
2x 3090 misadventures: Fallin off the bus.

### DISCUSSION THREAD

**Discussion by clort81** [noonghunna/club-3090#38] [General]
FELL OFF THE BUS!
-------------------------------

Dropped a 2nd 3090 onto a not-ideal motherboard.  Running linux kernel 6.12

2nd 3090 keeps "falling off the bus"

No way to reset the bus and bring-up card again, needs reboot.

Starting a discussion here in case we can collect other experiences with this kind of issue.

Possible Cause?

1. Bad 3090
2. Bad motherboard
3. A 3090 is inherently unstable on pcie1x slot
4. ?

Cheers!

**Comment by noonghunna** [comment]
Welcome [@clort81](https://github.com/clort81), and yeah — this is a known consumer-mobo class of issue. Worth talking through because we just shipped a feature that makes diagnosing it easier.

## Likely root cause (most common first)

A 3090 in a **PCIe x1 slot** is operating with ~1/16th the bandwidth of a normal x16 setup. Beyond just being slow, the bus is more sensitive to electrical/timing issues — the card "falling off the bus" usually means the link trains down to nothing or hits an AER (Advanced Error Reporting) error and the OS marks the device gone.

Common triggers, ordered by likelihood:

1. **PCIe Gen mismatch** — your minislot is probably a Gen 3 x1 slot wired off the chipset. The 3090 wants Gen 4 x16; if BIOS is set to "Auto" gen, it's negotiating across asymmetric capabilities. **Force the slot to Gen 3 x1** in BIOS (look for "PCIe Generation" or "Link Speed" per-slot).
2. **Insufficient power** — the x1 slot might be wired to a less robust power rail, especially on consumer mobos. If the card draws current spike during model load and the slot can't deliver, the card resets. **Lower the power limit** (`sudo nvidia-smi -pl 230 -i 1` to cap at 230W instead of factory 350-450W).
3. **PCIe ASPM (link power management)** — Linux + consumer mobos sometimes mis-handle ASPM on x1 slots. Try booting with kernel param `pcie_aspm=off`.
4. **Riser cable quality** — if you're using a riser to bring the x1 slot up to a full-size connector, cheap risers cause this exact pattern. Dell-style "official" mining risers tend to be more stable than generic ones.

## How to confirm what slot it's actually negotiating to

Pull latest club-3090 (`git pull origin master`) and run:

```bash
bash scripts/report.sh > my-rig.md
```

The GPU section now captures per-card PCIe lane width:

```
- **GPU 1:** NVIDIA GeForce RTX 3090 | 24576 MiB | driver ...
  - **PCIe:** x1 lanes negotiated (GPU max x16, Gen up to 4) | bus 00:08.0 ⚠ slot is narrower than GPU capability — affects load + all-reduce bandwidth
```

That confirms what the slot is actually negotiating to. If you see `x1 lanes negotiated` with the ⚠ flag, that's our case 1. If you see `x16 lanes negotiated` but the card still falls off the bus, it's an electrical/timing issue not a bandwidth one — different debug path.

## When it falls off the bus, capture these

Right after the card disappears (before reboot):

```bash
sudo dmesg -T | tail -50
sudo dmesg -T | grep -E "AER|pcieport|nvidia" | tail -20
nvidia-smi   # may hang or report Unable to query
lspci -vv -s <bus_id> | grep -E "LnkSta|LnkCap"
```

The `dmesg | grep AER` pattern is the gold standard — if there are correctable / uncorrectable errors logged, the bus is genuinely flaky. If dmesg shows nothing useful, it's likely a clean link drop (ASPM, power, cable).

## Mitigations if you can't move the card to a real slot

If the minislot is what you have, in order of effort:

1. **Single-card vLLM** — use the working slot, single-card config. `bash scripts/launch.sh --variant vllm/long-text` (180K + MTP, single 3090). Drops the second card from the equation entirely. ~50 narr / 67 code TPS.
2. **Don't use TP=2** — even if both cards are visible, NCCL all-reduce over a Gen 3 x1 slot would be brutal anyway. Per-stream throughput would be worse than single-card.
3. **Lower the unstable card's power limit** — `sudo nvidia-smi -pl 200 -i 1` reduces the chance of bus-reset from current spikes.

## What this is NOT (probably)

- Not a model issue (Qwen3.6-27B doesn't touch the bus differently than any other model)
- Not a vLLM/Genesis issue (this fires before the inference stack does anything)
- Not a kernel 6.12 issue specifically (this pattern is OS-version-agnostic)

It's a hardware/electrical/topology problem. The fix is at the BIOS/cabling/slot level, not the software stack.

If you can share the `report.sh` output + a `dmesg` capture next time it happens, we can probably narrow it further. Welcome to the club.

**Comment by clort81** [comment]
Thank you!

Just for clarity:
-------------------------
* Not using TP-2 or any tensor parallel.
* Not a vllm issue: Using Llama.cpp
* PSU issue unlikely: Problem persists at 200W setting for the 3090s: PSU is 1300W be quiet! Dark Power Pro 13 with seperate bars for each GPU.

Portions of the AI response invalid because:
---------------------------------------------------------------------
* RTX 3090 does not draw its bower from the PCIe bus
* PCIe speed: no documentation that a RTX3090 is unusable on PCIe3x1; little bandwidth required for inference with llama.cpp
* Linux flag:  Older documentation incorrectly stated that pcie_aspm=off would disable ASPM, intead it means "leaving firmware settings untouched." https://git.zx2c4.com/linux-rng/commit/?id=2e0239d47d75e08f9a39f8360610e490c39ea696

Will investigate:
------------------------------------------
* Stress testing each card with llama.cpp running each GPU single on primary PCIe slot
* BIOS-forcing PCIe to 3x1 and disabling ASPM in bios, so thanks for the response.
* BIOS-disabling ASPM

Further testing to follow when additional hardware arrives.

**Comment by noonghunna** [comment]
@clort81 — apologies for the late ack on this, and thanks for catching the factual errors in my prior reply. You're right on all three corrections:

1. **RTX 3090 power**: confirmed — 3090s have 2× 8-pin connectors (or 1× 12VHPWR on some FE/AIB variants); they draw essentially nothing through the PCIe slot itself. My "x1 slot can't deliver power spike" framing was wrong as stated. The slot delivers the standard 75W spec but the card barely uses it.

2. **PCIe x1 bandwidth for inference**: also right — for non-TP single-card llama.cpp inference, x1 is bandwidth-OK because the model is loaded to VRAM once at startup and the per-token traffic is small (logits + small KV ops). My implication that PCIe Gen mismatch causes "falling off the bus" was overgeneralized from TP-hot multi-card vLLM setups where allreduce traffic actually saturates the bus. For your single-card llama.cpp case it's a non-issue.

3. **`pcie_aspm=off` semantics**: thanks for the [git.zx2c4.com link](https://git.zx2c4.com/linux-rng/commit/?id=2e0239d47d75e08f9a39f8360610e490c39ea696). Updated my notes — current kernel docs are clearer that `pcie_aspm=off` means "don't touch firmware ASPM settings," not "force ASPM off." For actually disabling ASPM, you'd need `pcie_aspm.policy=performance` or BIOS-level disable.

I'll fold these corrections into our [docs/diagnostic-playbook](https://github.com/noonghunna/club-3090/tree/master/docs) the next time we touch hardware troubleshooting docs — your stress-testing-each-card-on-primary-slot approach is the right diagnostic methodology, and the corrected facts are a better baseline for future people hitting similar issues.

If your testing surfaces a specific BIOS setting / link state cause, would love a follow-up — "2nd 3090 on x1 minislot fell off the bus, fix was X" is exactly the kind of single-line lessons-learned that helps the next person down this path.

**Comment by clort81** [comment]
## Further observations:
* GPU single on primary PCIe slot much more stable, not 100% stable
* BIOS-forcing PCIe to 3x1, no improvement
* BIOS-disabling ASPM, no improvement

## Motherboard upgrade arrived with 2*x8 bifurcation, resizeable BAR bios, designed for 2xGPU.  

* Falling off bus definitely not caused by card running on PCIe x1.  
* Primary "reliable" card was also observed to fall off the bus once.

Experiments ongoing, possible stability win with lowered peak memory speed.

Looking more like stressed / damaged RTX 3090s.

**Reply by tomByrer** [reply]
> stressed / damaged RTX 3090s

What are the thermals?  Tried to repaste both?

If you really want to nerd out, you can also try to heatsink the memory better, blow extra fan directly on the card like I did, & add thermal putty to the capacitors for extra heat drainage.

**Comment by clort81** [comment]
My temps have been in the 60-80°C range during extended inference.  Busfalling was observed only after significant load.

I am happy to report that the sum total of all tweaks, new asspensive motherboard, has yielded stability!
---

Limiting max memory and gpu speed:
```
nvidia-smi -i [0,1] -lgc 300,1000
nvidia-smi -i [0,1] -lmc 800,7500
```
As well as *not* keeping nvtop running.  It maybe that program's polling is causing operational state change that exacerbates a memory timing problem.

Cards chugging along for two days now without a reboot!  9-months of GPU purgatory seem to be over.
